plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Debug/Debugger.hs

```diff
@@ -0,0 +1,267 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE ConstraintKinds #-}
+module UntypedPlutusCore.Evaluation.Machine.Cek.Debug.Debugger where
+
+import Annotation
+import UntypedPlutusCore
+import UntypedPlutusCore.Evaluation.Machine.Cek.Debug.Internal
+import Universe.Core
+
+import Data.RandomAccessList.Class qualified as Env
+import Data.Word64Array.Word8 hiding (Index, toList)
+import Control.Monad.State
+import Control.Monad.Trans.Free
+import Control.Lens hiding (Context)
+import Prettyprinter
+import Data.Sequence as Seq
+import Data.Coerce
+import GHC.Exts
+
+-- usually it is a line, but let's treat more generally as a srcspan
+newtype Breakpoint = Breakpoint { unBreakpoint :: SrcSpan }
+    deriving newtype (Show, Pretty)
+
+-- | The terms that we can debug
+type DTerm uni fun = Term NamedDeBruijn uni fun SrcSpans
+
+-- commands that the debugger can receive fromthe debug-client (tui,cli,test,etc)
+data Cmd
+  =
+  -- MAYBE: have a Load DTerm command; would require an initial invalid debugstate, i.e.  Maybe DebugState
+  Reload -- ^ reset back to start of initial term
+  -- small and big steps
+  | Step -- ^ step the cek once
+  | Next -- ^ step over aka next
+  | Finish -- ^ run to the end of current function
+  | ContinueBreak -- ^ continue to the end or until breakpoint reached
+  | ContinueIgnore -- ^ continue to the end by temporarily ignoring breakpoints
+  -- breakpoints
+  | BreakAdd Breakpoint -- ^ create a new breakpoint
+  | BreakDel Word -- ^ delete a breakpoint by index no.
+  | BreakDelAll -- ^ convenience function
+  | BreakList -- ^ list breakpoints
+  -- variables
+  | Print NamedDeBruijn
+  | PrintAll -- ^ all debruijn in the env
+  | Discharge NamedDeBruijn -- ^ dischargeCekValue
+  | DischargeAll -- ^ forall v in env . dischargeCekValue
+  | DischargeControl -- ^ dischargeCekValEnv
+  -- frames
+  | FrameList -- ^ aka Backtrace
+  | FrameFocus Word -- ^ switch to frame by index no.
+  -- Q: Is exit really needed?
+  -- Wouldn't it be simpler for the Client to just ignore/throw away the debugger computation?
+  | Exit -- ^ exit button or sth else control-d
+
+type ShowUniFun uni fun =
+    ( uni `Everywhere` Show
+    , Closed uni
+    , GShow uni
+    , Show fun
+    )
+
+data DebuggerState uni fun = DebuggerState
```

Comment:
Similar concern - different debugger UIs may need different state types.

Also, I'm not sure what "Debugger" means. There's a "CekState" and a "DebuggerState", but in the current implementation, I feel like Cek and Debugger are the same thing, which is the thing that supports all the `Cmd` above.